src/org/rascalmpl/values/parsetrees/SymbolFactory.java

```diff
@@ -337,31 +337,30 @@ else if (range.isFromTo()) {
 
 	private static IValue char2int(Char character) {
 		String s = ((Char.Lexical) character).getString();
-		if (s.startsWith("\\")) {
-			if (s.length() > 1 && java.lang.Character.isDigit(s.charAt(1))) { // octal escape
-				// TODO
-				throw new NotYetImplemented("octal escape sequence in character class types");
-			}
-			if (s.length() > 1 && s.charAt(1) == 'u') { // octal escape
-				// TODO
-				throw new NotYetImplemented("unicode escape sequence in character class types");
-			}
-			char cha = s.charAt(1);
+		if (s.matches("\\\\[auU][0-9A-F]+")) {
+			// ascii escape (a), utf16 escape (u) or utf24 escape (U)
+			return factory.integer(Integer.parseInt(s.substring(2), 16));
+		}
+		else if (s.startsWith("\\")) {
+			// builtin escape
+			int cha = s.codePointAt(1);
 			switch (cha) {
-			case 't': return factory.integer('\t');
-			case 'n': return factory.integer('\n');
-			case 'r': return factory.integer('\r');
-			case '\"' : return factory.integer('\"');
-			case '\'' : return factory.integer('\'');
-			case '-' : return factory.integer('-');
-			case '<' : return factory.integer('<');
-			case '>' : return factory.integer('>');
-			case '\\' : return factory.integer('\\');
+				case 't': return factory.integer('\t');
```

Comment:
do you know if the `'t'` is upgraded to an int, or if the 16bit highest bit of `cha` are dropped (aka downcast to a char)? If the latter, this can cause issues.

Reply:
I'm guessing the former because tihis code has been there for a while and it occurs often in whitespace classes like `[\ \t\n\r]`.

Reply:
https://docs.oracle.com/javase/specs/jls/se8/html/jls-5.html#jls-5.1.2

the char is widened to an int.

Reply:
I asked this becauase `cha` went from `char` to `int`. And there could be some high codepoints where only the high 16bits are different from `n`. Like:  [`0x2006E` aka 𠃦 (of which the lower bits are `0x006E` aka `n`)](https://www.compart.com/en/unicode/U+200E6)

Reply:
`int cha` is now the `int String::codepointAt` a given index. This method returns an `int`. And the literal java escapes we used are all below 7-bit ASCII. Their character values are safely widened to 32-bits by adding zeroes. Finally the Unicode escapes are meant to index into 24-bit Unicode codepoints directly (no surrogate encoding).

However, if you want to be smart you can take two low Unicode escapes and construct a high-surrogate/low surrogate pair. This will not work correctly. Probably you end up with two question marks in the console.

Reply:
The code we had before was broken for all the Unicode characters that have a high surrogate in char16 encoding. So this fix will repair that, but if you depended on the buggy behavior you now have a problem. That would be pretty unlikely since the high surrogates were chosen well in char16 encoding. They don't accidentally overlap with commonly used characters.